_config/requestprocessors.yml

```diff
@@ -1,11 +1,19 @@
 ---
 Name: requestprocessors
 ---
+SilverStripe\Control\Director:
+  middlewares:
```

Comment:
If we moved these to an injected service on top of director, rather than being a config of a list of service names, you could do away with the whole map-names-to-services loop you have every request. Something analogous to the RequestProcessor. This top level service itself could have the same method signature as HTTPMiddleware, but allow us to abstract the middleware handling itself out of Director. :D 


```yaml
SilverStripe\Injector\Injector:
  SilverStripe\Control\RequestMiddlewareProcessor:
    middleware:
      TrustedProxyMiddleware: '%$SilverStripe\Control\TrustedProxyMiddleware'
      AllowedHostsMiddleware: '%$SilverStripe\Control\AllowedHostsMiddleware'
      SessionMiddleware: 'SilverStripe\Control\SessionMiddleware'
      RequestProcessor: 'SilverStripe\Control\RequestProcessor'
      FlushMiddleware: '%$SilverStripe\Control\FlushMiddleware'
```

Then in Director just replace the  self::callWithhandler() with:

```php
$processor = Injector::inst()->get(RequestMiddlewareprocessor::class);
return $processor->process($request, $handler);
```

You could strip the middleware logic out of HTTPApplication into a trait, and have RequestMiddlewareprocessor use the same trait (same logic, some service). Overall less duplication of code and less code overall. :)